### PR TITLE
Apple Adventure

### PR DESCRIPTION
**Pull request by RandyGaul**
Getting things to work on Apple, at least a little bit

### CODE REVIEW

**Comment by RandyGaul** [comment]
Weirdly if _XOPEN_SOURCE is not defined near the top of the file, then the entire ucontext fallback fails. I think this define is affecting various includes on apple, so it's defined near the very tip top of everything

**Comment by RandyGaul** [comment]
I found three nice links with a ton of info on this topic.
- https://github.com/brimworks/lua/blob/master/src/lcoco.c
- http://lua-users.org/lists/lua-l/2007-04/msg00565.html
- https://gist.github.com/rsms/6238994

Search for the keyword `COCO_STACKADJUST`.

File: minicoro.h [hunk 1]
```diff
@@ -631,6 +631,9 @@ typedef struct _mco_ctxbuf {
 static void _mco_wrap_main(void) {
   __asm__ __volatile__ (
     "movq %r13, %rdi\n\t"
+#if defined(__APPLE__)
+    "sub $8, %rsp\n\t"
+#endif
```
Comment (edubart) [inline minicoro.h:636]: Offsetting here will probably make the dummy return address endup in the wrong place. The dummy return address assists debugging as the last function in the backtrace when you put a breakpoint inside a coroutine. The offset should be really on `_mco_makectx` to overcome this.

File: tests/mt-example.c [hunk 2]
```diff
@@ -1,16 +1,18 @@
 #define MINICORO_IMPL
 #include "minicoro.h"
-#define C89THREAD_IMPLEMENTATION
-#include "thirdparty/c89thread.h"
```
Comment (edubart) [inline tests/mt-example.c:4]: Actually I am trying to support the `c89thread` (from David Reid the author of `miniaudio`) so I don't wish to change the library here, I told him the issue there about the Mac OS compilation and he fixed, and the library has been updated in a recent commit.

**Review by edubart — CHANGES_REQUESTED** [review]

**Comment by edubart** [comment]
> I found three nice links with a ton of info on this topic.
> 
> * https://github.com/brimworks/lua/blob/master/src/lcoco.c
> * http://lua-users.org/lists/lua-l/2007-04/msg00565.html
> * https://gist.github.com/rsms/6238994
> 
> Search for the keyword `COCO_STACKADJUST`.

Thanks for helping out, researching this and getting minicoro running on Mac OS! From what we have read we need to offset then by 8 bytes and we should be fine on Mac.

**Comment by edubart** [comment]
Thanks @RandyGaul I went ahead and made the changes on my side in https://github.com/edubart/minicoro/commit/a8f3698d47f18e87fb1a416e98f5c8f035ebf395
so I will close this.

**Pull request closed without merging**